File: Leetcode July Challenge/C++/Partition Array into Disjoint Intervals.cpp
class Solution {
public:
    int partitionDisjoint(vector<int>& A) {
        
        int localMax = A[0], globalMax = A[0], pi=0;
        for(int i = 1; i < A.size(); i++){
            globalMax = max(globalMax, A[i]);
            if(localMax > A[i]){
                pi = i;
                localMax = globalMax;
            }
        }
        
        return pi+1;
//         int maxLeft = INT_MIN;
//         vector<int> minRight(A.size());
         
//         minRight[A.size()-1]=A.back();
        
//         for(int i=A.size()-2; i>=0; i--){
//             minRight[i]=min(A[i], minRight[i+1]);
//         }
            
        
//         for(int i = 0 ; i < A.size() - 1; i++){
//             maxLeft = max(maxLeft, A[i]);
//             if(maxLeft <= minRight[i+1]){
//                 return i+1;
//             }
//         }
//         return -1;
//         vector<int> maxLeft(A.size()), minRight(A.size());
//         maxLeft[0]=A[0]; 
//         minRight[A.size()-1]=A.back();
        
//         for(int i=1; i<A.size(); i++){
//             maxLeft[i]=max(A[i], maxLeft[i-1]);
//         }
//         for(int i=A.size()-2; i>=0; i--){
//             minRight[i]=min(A[i], minRight[i+1]);
//         }
            
        
//         for(int i = 0; i < A.size()-1; i++)
//             if(maxLeft[i] <= minRight[i+1])
//                 return i + 1;
        
//         return -1;
    }
};

File: Leetcode July Challenge/Python/Partition Array into Disjoint Intervals.py
def partitionDisjoint(self, A):
       maxleft = [A[0]]*len(A)
       for i in range(1,len(A)):
           maxleft[i] = max(maxleft[i-1], A[i])
           
       minright = [A[-1]]*len(A)
       for i in range(len(A)-2,-1,-1):
           minright[i] = min(minright[i+1], A[i]) 
       
       for i in range(len(A)-1):
           if maxleft[i] <= minright[i+1]:
               return i+1

def partitionDisjoint(self, A):
       minright = [A[-1]]*len(A)
       for i in range(len(A)-2,-1,-1):
           minright[i] = min(minright[i+1], A[i]) 

       maxleft = -float('inf')
       for i in range(len(A)-1):
           maxleft = max(maxleft, A[i])  
           if maxleft <= minright[i+1]:
               return i+1

def partitionDisjoint(self, A):
       disjoint = 0
       curmax = leftmax = A[0]
       for i,num in enumerate(A):
           curmax = max(curmax, num)
           if num < leftmax:
               leftmax = curmax
               disjoint = i
       return disjoint + 1  	
